doc/user: update What is Materialize? (#18888)
MaterializeInc/materialize: doc/user/content/overview/_index.md

    weight: 5
---

Materialize is a **streaming database** purpose-built for low-latency
applications. Data and software engineering teams use it to solve problems
where data must be processed at speeds and scales not possible in traditional
databases, but without the cost, complexity, and development time of most
streaming engines.

If you're looking to reduce the time it takes to refresh answers to common
queries, Materialize can make that happen! Rather than recalculating results
from scratch each time it's queried, or serving stale cached results,
Materialize continually ingests data and keeps results up to date for you
as **incrementally updated materialized views**.

## Features

Materialize combines all the things you love about databases and abstracts all
the hard parts that go into making a streaming engine horizontally scalable,
highly available and strongly consistent. To get up and running, all you need
to know is **SQL**.

### Incrementally updated views

> Materialized views are useful to avoid re-running heavy queries that are asked

Author: Most users will be familiar with the concept of materialized views, so this section can still be compacted a bit more.

    weight: 5
---

Materialize is a **streaming database** purpose-built for low-latency
applications. Data and software engineering teams use it to solve problems
where data must be processed at speeds and scales not possible in traditional
databases, but without the cost, complexity, and development time of most
streaming engines.

If you're looking to reduce the time it takes to refresh answers to common
queries, Materialize can make that happen! Rather than recalculating results
from scratch each time it's queried, or serving stale cached results,
Materialize continually ingests data and keeps results up to date for you
as **incrementally updated materialized views**.

<br>

> | ![streaming_analytics.gif](https://user-images.githubusercontent.com/23521087/233637537-4c2e5133-5a93-4d9d-9800-a0d4ba842c82.gif) |

Author: What do you think about using this Metabase dashboard as an example, @jseldess @tr0njavolta?

Author: Maybe we can come up with something more...official. This is just a GIF I had lying around! @ruf-io

Reviewer: What do you think about instead showing a graphic illustrating "incrementally updated materialized views" - so it'd be more utilitarian, two source tables, a SQL statement and an output, and showing how a change in input drives a quick change in output. Something like what we have on the homepage, but with source data tables added. 
![image](https://user-images.githubusercontent.com/11527560/234016350-ff10d5fb-acdc-473b-afe4-61e912d940ee.png)

I don't like using the BI dashboard UI here because, even though you say "as an example", since it's the overview page and it's the only visual people might incorrectly pigeonhole Materialize as being a thing for real-time analytics dashboards only.

to know is **SQL**.

### Incrementally updated views

Materialized views are useful to avoid re-running heavy queries that are asked
over and over against your database, and typically use caching to serve results
faster. In traditional databases, materialized views often come with a long
list of limitations and force you to make a compromise between the freshness of
the results, the cost of refreshing the view and the complexity of the SQL
statements you can use.

Materialize supports incrementally updated materialized views that are **always
fresh**, and allows a much broader set of SQL statements, like views over
multi-way joins with complex aggregations. How? Under the hood, its engine is
powered by [Timely](https://github.com/TimelyDataflow/timely-dataflow#timely-dataflow) and
[Differential Dataflow](https://github.com/timelydataflow/differential-dataflow#differential-dataflow),
backed by many years of research and optimized for this exact purpose. Best
yet: as a user, you don't really need to think about it — it just works™️.

### Standard SQL support

### PostgreSQL wire-compatibility

Every database needs a protocol to standardize communication with the outside

> world. Materialize uses the [PostgreSQL wire protocol](https://datastation.multiprocess.io/blog/2022-02-08-the-world-of-postgresql-wire-compatibility.html),

Author: Our [explainer](https://materialize.com/blog/postgres-compatibility/#1-postgresql-wire-compatibility) is very good, but it doesn't dig deeper into what it means to be wire-compatible with PostgreSQL. I find this blogpost from Phil Eaton very useful to understand the idea! 

This reminds me that we should  consider #8086 in #18508, @jseldess.

Reviewer 2: Thanks, @morsapaes. Added a comment here: https://github.com/MaterializeInc/materialize/issues/8086#issuecomment-1525770117.

    weight: 5
---

> Materialize is a **streaming database** purpose-built for low-latency

Reviewer 3: nice!

applications. You can use it to solve problems where data must be processed at
speeds and scales not possible in traditional databases, but without the cost,
complexity, and development time of most streaming engines.

If you need to reduce the time it takes to refresh queries that run frequently,
or trigger user-facing actions as soon as events happen, Materialize might be a
good fit for you! Rather than recalculate results from scratch each time it's
queried, or serve stale cached results, Materialize continually ingests data
and keeps results up to date for you as **incrementally updated materialized
views**.

<br>

| ![streaming_analytics.gif](https://user-images.githubusercontent.com/23521087/233637537-4c2e5133-5a93-4d9d-9800-a0d4ba842c82.gif) |
|:---------------------------------------------------------------------------------------------------------------------------------:|

> | *As an example, you can use Materialize to keep your dashboards automatically up to date.*                                                           |

Reviewer 3: I like having a visual example on this page! it might be nice to make a custom visualization for this, sort of like the ones we have on the homepage of the website?

Reviewer 3: not required! just a thought.

Author: I agree!

## Features

Materialize combines all the desirable features of SQL databases and abstracts
away all the hard parts required to build a streaming engine horizontally
scalable, highly available and strongly consistent. To get up and running, all
you need to know is **SQL**.

> ### Incrementally updated views

Reviewer 3: thought: we should probably land on a single way of describing this?

incrementally-maintained materialized views
incrementally-updated views
etc.

Author: Good point! We seem to use a mix of both in the documentation. I find "incrementally updated" more immediate to understand.

Materialized views are useful to avoid re-running heavy queries that run
frequently against a database, and typically use caching to serve results
faster. In traditional databases, materialized views often come with a long
list of limitations and force you to make a compromise between the freshness of
the results, the cost of refreshing the view and the complexity of the SQL
statements you can use.

> Materialize supports incrementally updated materialized views that are **always

Reviewer 3: nit: the two parts of this sentence feel a little unrelated atm. maybe...

Materialize supports incrementally updated materialized views that are **always fresh**, even when the SQL contains multi-way joins and complex aggregations.

fresh**, and allows a much broader set of SQL statements, like views over

> multi-way joins with complex aggregations. How? Under the hood, its engine is

Reviewer 3: nit: the multiple `and`s in this sentence feel a little clunky. maybe...

How? Under the hood, Materialize is powered by Timely and Differential Dataflow, systems backed by many years of research and optimized for this exact purpose.

powered by [Timely](https://github.com/TimelyDataflow/timely-dataflow#timely-dataflow) and
[Differential Dataflow](https://github.com/timelydataflow/differential-dataflow#differential-dataflow),
backed by many years of research and optimized for this exact purpose.

### Standard SQL support

You interact with Materialize like you would with any other database: using
standard SQL. We implement a large fraction of PostgreSQL, and are actively
working on supporting more built-in PostgreSQL functions, operators and data
types, as well as extending our `pg_catalog` support.

> With Materialize, you can build complex analytical workloads using **any type of

Reviewer 3: should we link to our join documentation here?

Author: Ah, good catch!

> complexity, and development time of most streaming engines.

Author: What I wanted to express is "Materialize avoids the complexity of adopting a stream processing framework when your use case outgrows traditional databases". So closer to your suggestion, @jseldess. Using "but" makes the sentence clearer to me, but then again I'm not a native speaker!

> views**.

Author: "maintains results up to date"?

> Materialized views are useful to avoid re-running heavy queries that run
> frequently against a database, and typically use caching to serve results
> faster. In traditional databases, materialized views often come with a long
> list of limitations and force you to make a compromise between the freshness of
> the results, the cost of refreshing the view and the complexity of the SQL
> statements you can use.

Reviewer 4: ```suggestion

Creating materialized views avoids re-running frequent heavy queries against a
database because they cache queries to deliver faster results. In traditional
databases, materialized views limit the complexity of SQL statements you can
use and require time and resources to return the most up-to-date results.
```

> aggregations. How? Under the hood, Materialize is powered by [Timely](https://github.com/TimelyDataflow/timely-dataflow#timely-dataflow)
> and [Differential Dataflow](https://github.com/timelydataflow/differential-dataflow#differential-dataflow)
> — data processing frameworks backed by many years of research and optimized for
> this exact purpose.

Reviewer 4: ```suggestion
aggregations. How? Materialize runs on [Timely](https://github.com/TimelyDataflow/timely-dataflow#timely-dataflow)
and [Differential Dataflow](https://github.com/timelydataflow/differential-dataflow#differential-dataflow)
— data processing frameworks backed by many years of research and optimized for
this exact purpose.
```

> standard SQL. We implement a large fraction of PostgreSQL, and are actively

Author: Directly lifted from: https://github.com/MaterializeInc/materialize#about

This means that we follow the PostgreSQL SQL implementation wrt [conformance to the SQL standard](https://www.postgresql.org/docs/current/features.html). I'll try to phrase it in a clearer way.

which allows it to integrate out-of-the-box with most SQL clients and other
tools in the data ecosystem that support PostgreSQL — including [dbt](/integrations/dbt/).

Don't see the a tool that you’d like to use with Materialize listed under [Tools and integrations](/integrations/)? Let us know
by submitting a [feature request](https://github.com/MaterializeInc/materialize/issues/new?assignees=&labels=A-integration&template=02-feature.yml)!

### Strong consistency guarantees

By default, Materialize provides the highest level of transaction isolation:
[**strict serializability**](https://jepsen.io/consistency/models/strict-serializable).
This means that it presents as if it were a single process, despite spanning a
large number of threads, processes and machines. You get to avoid common
pitfalls of eventual consistency, dual writes, and other details that affect
the correctness of your results but you shouldn't need to think about. You can
[adjust the transaction isolation level](/overview/isolation-level/) depending on
your consistency and performance requirements.

> <hr>

Author: This is missing a wrap-up with a CTA, but I'll leave that and bridging between sections to your best judgement, @jseldess! If we think it's valuable, I can add a very brief section for use cases later, using the following structure:

```
## Use cases

### Real-Time analytics

### Automation and alerting

### Segmentation and personalization
```
